### DISCUSSION TITLE
How to use Nuxt 3 programmatically?

### DISCUSSION THREAD

**Discussion by glopezep** [nuxt/nuxt#16286] [Questions]
.

**Comment by wobsoriano** [comment]
More info please

https://github.com/nuxt/framework/discussions/1451

**Reply by hnviradiya** [reply]
Nuxt3 exported build is not working with import syntax

**Comment by danielroe** [comment]
See https://v3.nuxtjs.org/docs/advanced/kit.

**Reply by ValeriyDP** [reply]
I think this is correct link https://nuxt.com/docs/api/advanced/kit/

**Comment by hnviradiya** [comment]
I am getting below error (which is same as using with kit or without kit)

```
Uncaught ReferenceError ReferenceError: exports is not defined in ES module scope
This file is being treated as an ES module because it has a '.js' file extension and '[c:\Users\hardik\Documents\GitLab\myproject\package.json]()' contains "type": "module". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.
    at <anonymous> ([c:\Users\hardik\Documents\GitLab\myproject\dist\src\main.js:2:23]())
    at run (node:internal/modules/esm/module_job:195:25)
    --- async function ---
    at runMainESM (node:internal/modules/run_main:51:21)
    at executeUserEntryPoint (node:internal/modules/run_main:74:5)
    at <anonymous> (node:internal/main/run_main_module:17:47)
```

**Comment by hnviradiya** [comment]
Following is what I am using (have commented nuxt3 import which I had used previously). 

```
import { NestFactory } from '@nestjs/core';
// import { build, loadNuxt } from 'nuxt3';
import { AppModule } from './app.module';
import { loadNuxt, buildNuxt } from '@nuxt/kit'

async function bootstrap() {
  const app = await NestFactory.create(AppModule);

  // Check if we need to run Nuxt in development mode
  const isDev = process.env.NODE_ENV !== 'production'

  // Get a ready to use Nuxt instance
  const nuxt = await loadNuxt({ rootDir: '/client/' })

  // Enable live build & reloading on dev
  if (isDev) {
    buildNuxt(nuxt)
  }

  await app.listen(3000);
}
bootstrap();
```

**Reply by danielroe** [reply]
The issue is not with `@nuxt/kit`. See https://stackblitz.com/edit/node-ezqqwt.

**Reply by hnviradiya** [reply]
Okk. Now I am using typescript which is unable to use this mjs import because it exports to js instead of mjs. :(

See below I have edited from your stackblitz.

https://stackblitz.com/edit/node-n4te8a

Seems typescript is not supporting mjs import (old post, but not sure if that is still supported today. and it is also mentioned that there is no workaround as well to this except for changing file types to mjs manually)

https://medium.com/@iamstan/typescript-es-modules-micheal-jackson-2040216be793

How do you people use this with typescript? Isn't nuxt 3 developed in typescript only?

**Reply by danielroe** [reply]
`.js` files can't import from ESM-only packages, as they run by default in CJS context. You can add `type: module` to your package.json. See https://stackblitz.com/edit/node-nrhc4. You might also find it helpful to read through [this issue](https://github.com/Microsoft/TypeScript/issues/18442) and [this documentation](https://nodejs.org/api/esm.html).

**Reply by dathoangnd** [reply]
> Following is what I am using (have commented nuxt3 import which I had used previously).
> 
> ```
> import { NestFactory } from '@nestjs/core';
> // import { build, loadNuxt } from 'nuxt3';
> import { AppModule } from './app.module';
> import { loadNuxt, buildNuxt } from '@nuxt/kit'
> 
> async function bootstrap() {
>   const app = await NestFactory.create(AppModule);
> 
>   // Check if we need to run Nuxt in development mode
>   const isDev = process.env.NODE_ENV !== 'production'
> 
>   // Get a ready to use Nuxt instance
>   const nuxt = await loadNuxt({ rootDir: '/client/' })
> 
>   // Enable live build & reloading on dev
>   if (isDev) {
>     buildNuxt(nuxt)
>   }
> 
>   await app.listen(3000);
> }
> bootstrap();
> ```

Does Nuxt server run? Or it simply build Nuxt?
I use the same code but can not access Nuxt, visit localhost:3000, it returns 404 as an Express route. How can we access Nuxt?

**Reply by hnviradiya** [reply]
@dathoangnd for that we need to use render function which was available in nuxt 2, but unable to find in nuxt 3. @danielroe is there any method in any other package? I tried finding in kit but couldn't see any similar way there.

**Reply by dathoangnd** [reply]
Yes, I used to use the render function in Nuxt 2, but can not find a similar thing in Nuxt 3, so can not upgrade the stack :(
